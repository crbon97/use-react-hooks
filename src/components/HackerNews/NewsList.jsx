import "./style.scss";

import React, { useEffect, useState } from "react";

import News from "./News";
import TextInput from "../MyForm/TextInput";
import axios from "axios";

const API_URL = "https://hn.algolia.com/api/v1";

const initialState = {
  query: "",
  items: [],
  loading: false,
  error: ""
};
const NewsList = props => {
  const [state, setState] = useState(initialState);
  const { query, items, loading, error } = state;
  const handleChangeText = e => {
    const { value } = e.currentTarget;
    setState(prevState => ({ ...prevState, query: value }));
  };
  useEffect(() => {
    setState(prevState => ({ ...prevState, loading: true, erorr: "" }));
    axios
      .get(`${API_URL}/search?query=${query}`)
      .then(res => {
        console.log(res);
        const { hits = [] } = res.data || {};
        setState(prevState => ({
          ...prevState,
          items: hits,
          loading: false,
          error: ""
        }));
      })
      .catch(e =>
        setState(prevState => ({
          ...prevState,
          error: e.message,
          loading: false,
          items: []
        }))
      );
    // Can return a callback to clean in componentWillUnmount phase
  }, [query]);
  return (
    <div className="hacker-news-container">
      <span className="field-label">Search:</span>
      <TextInput
        className="field-text"
        value={query}
        onChangeText={handleChangeText}
      />
      {loading && <p data-testid="loading">Loading....</p>}
      {!loading && !!error && <p data-testid="error">Error: {error}</p>}
      {!loading && items.length > 0 && (
        // TODO: We can paginate with parameter page=x in the URL
        // Follow: https://hn.algolia.com/api for more information
        <ol data-testid="result">
          <span>Results:</span>
          {items
            .filter(item => !!item.url)
            .map(item => (
              <News key={item.objectID} {...item} />
            ))}
        </ol>
      )}
    </div>
  );
};

export default NewsList;
