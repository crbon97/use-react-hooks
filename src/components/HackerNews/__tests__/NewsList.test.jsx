import { act, fireEvent, render, waitForElement } from "react-testing-library";

// import MockAdapter from "axios-mock-adapter";
import NewsList from "../NewsList";
import React from "react";
import axiosMock from "axios";
import renderer from "react-test-renderer";

jest.mock("axios");
const hits = [
  {
    title: "React 16",
    url: "https://facebook.github.io/react/blog/2017/09/26/react-v16.0.html",
    objectID: "1"
  }
];

describe("Test News component with `react-test-renderer`", () => {
  test("News should render without error", () => {
    const newsListWrapper = renderer.create(<NewsList />);
    let tree = newsListWrapper.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
describe("Test News component with `react-testing-library`", () => {
  test("News should render with the right props", async () => {
    //https://www.leighhalliday.com/async-axios-react-testing-library
    let { getByTestId, rerender } = render(<NewsList />);
    axiosMock.get.mockResolvedValue({ data: { hits } });
    act(() => {
      ({ getByTestId } = rerender(<NewsList />));
    });
    expect(getByTestId("loading").textContent).toContain("Loading....");
    const resolvedElement = await waitForElement(() => getByTestId("result"));
    expect(resolvedElement).toBeTruthy();
  });
});
