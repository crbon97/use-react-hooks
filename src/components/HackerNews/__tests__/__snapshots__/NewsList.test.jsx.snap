// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test News component with \`react-test-renderer\` News should render without error 1`] = `
<div
  className="hacker-news-container"
>
  <span
    className="field-label"
  >
    Search:
  </span>
  <input
    className="field-text"
    onChange={[Function]}
    type="text"
    value=""
  />
</div>
`;
